### DISCUSSION TITLE
Outreachy: WCAG 2.2 audit of wagtail.org – Dec 2023

### DISCUSSION THREAD

**Discussion by albinazs** [wagtail/wagtail#11305] [Development]
👋 opening a separate discussion of an upcoming Outreachy project: WCAG audit of wagtail.org. We’re looking for feedback/ suggestions on how we plan the audit, the specifics of the testing, and how we share reports / tackle the issues.

## Purpose

Create a backlog of accessibility improvements for [wagtail.org](https://wagtail.org/)

### Target standard

We can start with the most well-established [WCAG 2.2 AA-level](https://www.w3.org/TR/WCAG22/) (Recommendation as of October 2023), and some of AAA-level guidelines (if they seem to be easily solvable), as well as best practice recommendations (such as WHCM, print styles, and readability) as we see fit. We could also attempt to cover a few concepts from draft WCAG 3.0, such as color contrast. 

### Scope

Ideally, all types of pages on prod + possibly 3rd party forms.

### Methodology

Possible options to consider:
- Single-page semi-automated checks with [Accessibility insights](https://accessibilityinsights.io/), Lighthouse, [Sa11y](https://sa11y.netlify.app/) or similar
- Manual keyboard navigation checks
- Manual colour contrast check for non-text UI elements and images
- Manual zoom & magnification testing with Chrome &OS Zoom
- Full screen reader testing with NVDA in Firefox ESR
- Mobile touch interaction testing with Android Chrome
- Screenshot-taking & visual review for tab order
- Speech recognition: Dragon, Voice Control
- Contrast themes (Windows, or emulation in Chrome)
- Print stylesheets

### Reporting

Could be [our standard template](https://docs.google.com/spreadsheets/d/1kWb6f9YlT-UA2TC4ioAH40Y5T18hd_Hbt_ncTamabW4/edit?usp=sharing)

**Comment by ShubhamOulkar** [comment]
Hi, @albinazs  I would like to participate in the Accessibility audit testing.

**Reply by thibaudcolas** [reply]
Hi @ShubhamOulkar 👋 We’d love to have more people involved with the testing if you’re up for it. Here are three pages we haven’t been able to include in the audit so far – would you be able to audit those, with the methodology we outlined?

- https://wagtail.org/wagtail-vs-wordpress/
- https://wagtail.org/who-uses-wagtail/
- https://wagtail.org/packages/

I’d recommend focusing on the main content and ignoring the pages’ main nav, header, and footer areas, as we’ve audited them on other similar pages. Please report your findings in this thread.

**Reply by ShubhamOulkar** [reply]
Hi @thibaudcolas, I did an accessibility audit on page [wagtail-vs-wordpress](https://wagtail.org/wagtail-vs-wordpress/).
[Report (WIP)](https://docs.google.com/spreadsheets/d/1XBPWf-yuRkOlRJddOlWQ8Q2k0oFPkkn0CcRv503KeaA/edit?usp=sharing)

I found major issues reported in the report. 

`Q.1`  Why does this page have a signup form at the middle of the page? This form has issues. (footer form is OK)
`Q.2` I want to know from you that color contrast for links on hover does not pass AAA standards in `Dark mode`. Is it a problem with dark mode or proper color is not chosen? 
 
Kindly report my mistakes in this thread. So that I can improve on accessibility skills. Or you can add comments in Doc file.

**Reply by thibaudcolas** [reply]
@ShubhamOulkar thank you! I have integrated your findings in the main audit, and added comments for the points I’m not clear on. Note "AAA" isn’t in scope as far as colors. For readability, we’d need more actionable findings. And "Major" has quite a specific meaning for us, please see the About sheet in the spreadsheet.

- Why does this page have a signup form at the middle of the page? I’m not sure :) I believe the idea is to not wait until the very end to call people reading this to action
- We don’t do AAA-level color contrast for colors, so this is a non-issue.

**Comment by thibaudcolas** [comment]
Progress to date by @shyusu4:

- ✅ Single-page semi-automated checks with [Accessibility insights](https://accessibilityinsights.io/), Lighthouse, [Sa11y](https://sa11y.netlify.app/) or similar
- ✅ Manual keyboard navigation checks
- ✅ Manual colour contrast check for non-text UI elements and images
- ✅ Manual zoom & magnification testing with Chrome &OS Zoom
- ✅ Full screen reader testing with NVDA in Firefox ESR
- ✅ Mobile touch interaction testing with Android Chrome
- ✅ Screenshot-taking & visual review for tab order
- 🔴 Speech recognition: Dragon, Voice Control
- ⌛Contrast themes (Windows, or emulation in Chrome)
- ⌛Print stylesheets

We’ve decided to skip speech recognition tests for now (unsure which options to use on Windows)

**Comment by thibaudcolas** [comment]
Just a quick note to say @shyusu4 has completed her [wagtail.org accessibility audit](https://docs.google.com/spreadsheets/d/1kWb6f9YlT-UA2TC4ioAH40Y5T18hd_Hbt_ncTamabW4/edit), and we’re now switching to fixing of the issues.

See for example:

- https://github.com/wagtail/wagtail.org/issues/456
- https://github.com/wagtail/wagtail.org/issues/454

And a first pull request (congrats @shyusu4!): https://github.com/wagtail/wagtail.org/pull/455

**Comment by thibaudcolas** [comment]
Sharing @shyusu4’s executive summary for the audit for future reference:

## **WCAG 2.2 Accessibility Audit: Wagtail.org**

This executive summary provides an overview of the purpose, scope, methodology, and findings of the accessibility audit conducted on[ wagtail.org](https://wagtail.org/).

The purpose of the audit was to identify any accessibility issues on the website in compliance with Web Content Accessibility Guidelines (WCAG) V2.2 Level AA.

### **Scope**

The audit focused on selected public-facing web pages of the site, including the [Home](https://wagtail.org/),[ Features](https://wagtail.org/features/),[ Roadmap](https://wagtail.org/roadmap/),[ Services](https://wagtail.org/services/),[ Blog](https://wagtail.org/blog/) pages and a sample blog post, "[Fix a common Wagtail UX issue in two lines](https://wagtail.org/blog/fix-a-common-wagtail-ux-issue-in-two-lines/)." Pages located on sub-domains were excluded.

### **Methodology**

The site was evaluated using Google Chrome on Windows and Android, and Safari via BrowserStack emulation on Mac and iPhone. The audit methodology incorporated a combination of semi-automated and manual checks to assess accessibility, including

#### Semi-automated checks

* Accessibility insights
* Sa11y
* Lighthouse

#### Manual checks

* Keyboard navigation
* Chrome full-page and font-size zoom
* Windows High Contrast Mode
* NVDA on Windows
* TalkBack on Android
* VoiceOver on IOS and MacOS
* Print Stylesheets

### **Main findings**

The audit has highlighted several key findings, including

Discrepancies between screen reader software (TalkBack vs. NVDA and VoiceOver)

For instance, in header navigation submenus, links within the toggle menu are being announced by Android TalkBack with "minus" due to the presence of a _<span class="u-sr-only">-</span>_ element.

Android TalkBack also demonstrates multiple instances of variations in announcing elements, particularly links, when compared to other screen readers. This inconsistency may lead to a fragmented user experience for individuals relying on different assistive technologies.

The cookie policy modal requires improvement in overall accessibility for screen reader users. There are navigation challenges, since screen reader users typically navigate using headings or links for efficiency, but the modal lacks a heading and relies on non-descriptive link text ("find out more"). 

Throughout the site, there is a common case of decorative shapes covering text in section heading blocks which hinders readability.

### **Conclusion**

The overall state of accessibility on [wagtail.org](https://wagtail.org/) is considered to be in good shape. The majority of problems noted are categorized as minor. Addressing these concerns will ensure the website maintains a high level of accessibility for all users.

**Comment by thibaudcolas** [comment]
Sharing the audit executive summary by @shyusu4 for future reference:

## **WCAG 2.2 Accessibility Audit: Wagtail.org**

This executive summary provides an overview of the purpose, scope, methodology, and findings of the accessibility audit conducted on[ wagtail.org](https://wagtail.org/).

The purpose of the audit was to identify any accessibility issues on the website in compliance with Web Content Accessibility Guidelines (WCAG) V2.2 Level AA.

View the full audit findings: [Wagtail.org | 2024 accessibility audit spreadsheet](https://docs.google.com/spreadsheets/d/1kWb6f9YlT-UA2TC4ioAH40Y5T18hd_Hbt_ncTamabW4/edit)

### **Scope**

The audit focused on selected public-facing web pages of the site, including the [Home](https://wagtail.org/),[ Features](https://wagtail.org/features/),[ Roadmap](https://wagtail.org/roadmap/),[ Services](https://wagtail.org/services/),[ Blog](https://wagtail.org/blog/) pages and a sample blog post, "[Fix a common Wagtail UX issue in two lines](https://wagtail.org/blog/fix-a-common-wagtail-ux-issue-in-two-lines/)." Pages located on sub-domains were excluded.

### **Methodology**

The site was evaluated using Google Chrome on Windows and Android, and Safari via BrowserStack emulation on Mac and iPhone. The audit methodology incorporated a combination of semi-automated and manual checks to assess accessibility, including

#### Semi-automated checks

* Accessibility insights
* Sa11y
* Lighthouse

#### Manual checks

* Keyboard navigation
* Chrome full-page and font-size zoom
* Windows High Contrast Mode
* NVDA on Windows
* TalkBack on Android
* VoiceOver on IOS and MacOS
* Print Stylesheets

### **Main findings**

The audit has highlighted several key findings, including

Discrepancies between screen reader software (TalkBack vs. NVDA and VoiceOver)

For instance, in header navigation submenus, links within the toggle menu are being announced by Android TalkBack with "minus" due to the presence of a _<span class="u-sr-only">-</span>_ element.

Android TalkBack also demonstrates multiple instances of variations in announcing elements, particularly links, when compared to other screen readers. This inconsistency may lead to a fragmented user experience for individuals relying on different assistive technologies.

The cookie policy modal requires improvement in overall accessibility for screen reader users. There are navigation challenges, since screen reader users typically navigate using headings or links for efficiency, but the modal lacks a heading and relies on non-descriptive link text ("find out more"). 

Throughout the site, there is a common case of decorative shapes covering text in section heading blocks which hinders readability.

### Statistics

Total issues reported: 52 (47 by @shyusu4, 5 by @ShubhamOulkar)

#### Issues reported by testing technique

| Testing technique                                      | Number of issues |
| ----------------------------------------- | ---------------- |
| Other manual testing (Chrome)             | 20               |
| High-contrast mode (Windows, Chrome)      | 8                |
| Talkback (Android Chrome)                 | 5                |
| Print Stylesheets (Chrome)                | 5                |
| Sa11y (Chrome)                            | 4                |
| Mobile touch interaction (Android Chrome) | 3                |
| NVDA (Windows Chrome)                     | 3                |
| Accessibility Insights (Axe, Chrome)      | 2                |
| Full-page zoom (Chrome)                   | 1                |
| Keyboard navigation (Chrome)              | 1                |


#### Issues reported by severity

| Issue Severity     | Issues found | Explanation                                                                                  | Recommendation                                 |
| ------------------ | ------------ | -------------------------------------------------------------------------------------------- | ---------------------------------------------- |
| 1 Blocker          | 2            | Critical issue. Certain users cannot perceive or operate this, with no workaround available. | Fix first                                      |
| 2 Major            | 3            | Certain users are very likely to have difficulty perceiving or operating this.               | Fix second                                     |
| 3 Confusing        | 19           | Understanding or operating this may be difficult for some users.                             | Fix if simple, otherwise evaluate alternatives |
| 4 Minor            | 21           | Degraded user experience for some users. This slows down use of the application.             | Fix if simple, otherwise evaluate alternatives |
| 5 Validation issue | 0            | This element is not implemented to spec, but is not confirmed to have known issues.          | Evaluate case by case                          |
| 6 FYI only         | 7            | Notes and suggestions; no action needed from an accessibility perspective.                   | Evaluate case by case                          |

#### Issues reported by accessibility guideline

| Guideline                          | Issues found |
| ---------------------------------- | ------------ |
| Non-standard, established practice | 31           |
| 1.3.1 Info and Relationships       | 3            |
| 1.4.11 Non-text Contrast           | 3            |
| 1.4.3 Contrast (Minimum)           | 2            |
| 1.4.4 Resize Text                  | 2            |
| 1.4.10 Reflow                      | 2            |
| 2.5.8 Target size (Minimum)        | 2            |
| 3.2.4 Consistent Identification    | 2            |
| 1.3.4 Orientation                  | 1            |
| 1.4.5 Images of Text               | 1            |
| 1.1.1 Non-text Content             | 1            |
| 2.4.6 Headings and Labels          | 1            |
| 2.4.13 Focus Appearance            | 1            |

### **Conclusion**

The overall state of accessibility on [wagtail.org](https://wagtail.org/) is considered to be in good shape. The majority of problems noted are categorized as minor. Addressing these concerns will ensure the website maintains a high level of accessibility for all users.

**Comment by thibaudcolas** [comment]
@shyusu4 is working on a report of her own for the project but in the meantime – here is the project report information we’ve submitted to Outreachy.

## **Summary of the project goals completed**

Primary goals:

* **Accessibility audit planning**: 100% completed, selected 6 representative pages to test with 11 different types of testing
* **Accessibility testing**: 100% completed – found 47 different issues. See [https://github.com/wagtail/wagtail/discussions/11305](https://github.com/wagtail/wagtail/discussions/11305) for full details of audit
* **Accessibility audit report**: 85% completed – executive summary available in GitHub ([https://github.com/wagtail/wagtail/discussions/11305#discussioncomment-8383881](https://github.com/wagtail/wagtail/discussions/11305#discussioncomment-8383881)), spreadsheet fully completed, 21 issues still to be moved to the GitHub backlog
* **Implement accessibility improvements**: completed fixes for 18 out of 20 issues earmarked for fixing within the scope of the project

Secondary goals:

* **Exploring accessibility best practices**: done for print stylesheets, WHCM, APCA color contrast, hgroup element
* **Sharing best practices with the community**: completed: sharing audit methodology & findings, discussions within accessibility team meetings. Work in progress: final project blog post or blog post series to report on the work, participation to upcoming community webinar

Stretch goals:

* **Auditing other sites**: completed a headline review of [https://areweheadlessyet.wagtail.org/](https://areweheadlessyet.wagtail.org/)
* **Automated tests**: postponed
* **Adding features in Wagtail**: postponed
* **Implement accessibility improvements on other sites**: fixed 4 issues on [https://areweheadlessyet.wagtail.org/](https://areweheadlessyet.wagtail.org/), 2 more issues scheduled to fix within the last week

## **Impact the project will have on the open source community**

Thanks to this project, we have an excellent sense of how accessible our main website is – and a clear path towards fixing all known issues. We’re aware of where there are potential blockers / major issues for us to prioritise, and can use this information to inform future project plans.

With all of this work being done in the open, we also have a solid case study of what it looks like to improve the accessibility of a website built with our software, and a good sense of what issues to look out for in other projects.

Finally, with Sherry being keen to write a detailed project report and to appear on our community webinar at the end of the project, this will further build up our profile as being an accessibility-focused project.

## **What your intern learned during the internship**

* Accessibility audit planning
* Accessibility testing
* Working within an accessibility team
* WCAG 2.2
* Screen reader testing: Talkback, NVDA
* WHCM / Contrast themes
* Print stylesheets
* Technical writing
* Accessibility audit reporting
* Stakeholder presentations
* Django / Python tooling
* Node front-end tooling
* Git & GitHub
* Django Templates
* ARIA
* Next.js, React, TypeScript
* CSS modules
